File: Core/Entities/Product.cs
using System;
using Core.Interfaces;

namespace Core.Entities;

public class Product : BaseEntity, IDtoConvertible
{

    public required string Name { get; set; }
    public required string Description { get; set; }
    public decimal Price { get; set; }
    public required string PictureUrl { get; set; }
    public int ProductTypeId { get; set; }
    public ProductType ProductType { get; set; } = null!;
    public required string Brand { get; set; }
    public int QuantityInStock { get; set; }
    public bool IsSuggested { get; set; }
    public bool IsBestSelling { get; set; }
    public bool IsBestReviewed { get; set; }
    public float? AdminRating { get; set; }
    public ICollection<Favourite> Favourites { get; set; } = [];
    public ICollection<Review> Reviews { get; set; } = [];
    public ICollection<Discount>? Discounts { get; set; } = [];
    public ICollection<ProductImage> Images { get; set; } = [];
}

File: client/src/app/shared/models/product.ts
import { ProductReview } from './product-review.model';
import { ProductTypeDto } from './product-type.model';

export type ProductImage = {
  id: number;
  url: string;
  thumbnailUrl?: string;
  displayOrder: number;
  isPrimary: boolean;
  altText: string;
}

export type Product = {
  id: number;
  name: string;
  description: string;
  price: number;
  originalPrice?: number;
  discountPercentage?: number;
  hasActiveDiscount?: boolean;
  discountName?: string;
  pictureUrl: string;
  type: string; // Keep for backward compatibility in display
  productTypeId?: number; // Optional for backward compatibility
  productType?: ProductTypeDto; // New field for structured product type
  brand: string;
  quantityInStock: number;
  isFavourite: boolean;
  rating?: number;
  reviews?: ProductReview[];
  images?: ProductImage[];
  reviewsCount?: number;
}

export type CreateProductRequest = {
  name: string;
  description: string;
  price: number;
  pictureUrl: string;
  productTypeId: number;
  brand: string;
  quantityInStock: number;
}